yamux/src/lib.rs

```diff
@@ -95,70 +76,89 @@ pub enum WindowUpdateMode {
+            // TODO
+            max_num_streams: 512,
             read_after_close: true,
             split_send_size: DEFAULT_SPLIT_SEND_SIZE,
         }
     }
 }
 
 impl Config {
-    /// Set the receive window per stream (must be >= 256 KiB).
+    /// Set the maximum receive window per stream .
+    ///
+    /// Must be `>= 256 KiB`.
+    ///
+    /// The window of a stream starts at 256 KiB and is increased (auto-tuned) based on the
+    /// connection's round-trip time and the stream's bandwidth (striving for the
+    /// bandwidth-delay-product).
+    ///
+    /// Set to `None` to disable the per stream maximum receive window.
     ///
     /// # Panics
     ///
     /// If the given receive window is < 256 KiB.
-    pub fn set_receive_window(&mut self, n: u32) -> &mut Self {
-        assert!(n >= DEFAULT_CREDIT);
-        self.receive_window = n;
+    pub fn set_max_stream_receive_window(&mut self, n: Option<u32>) -> &mut Self {
+        self.max_stream_receive_window = n;
+        self.check();
         self
     }
 
-    /// Set the max. buffer size per stream.
-    pub fn set_max_buffer_size(&mut self, n: usize) -> &mut Self {
-        self.max_buffer_size = n;
+    // TODO: Is a usize really needed here?
+    // TODO: Should this be an option?
+    /// Set the maximum sum of all stream receive windows per connection.
+    ///
+    /// Must be `>= 256 KiB * max_num_streams` to allow each stream the Yamux default window size.
+    pub fn set_max_connection_receive_window(&mut self, n: usize) -> &mut Self {
+        self.max_connection_receive_window = n;
+        self.check();
         self
     }
 
-    /// Set the max. number of streams.
+    /// Set the max. number of streams per connection.
     pub fn set_max_num_streams(&mut self, n: usize) -> &mut Self {
         self.max_num_streams = n;
-        self
-    }
-
-    /// Set the window update mode to use.
-    pub fn set_window_update_mode(&mut self, m: WindowUpdateMode) -> &mut Self {
-        self.window_update_mode = m;
+        self.check();
         self
     }
 
     /// Allow or disallow streams to read from buffered data after
     /// the connection has been closed.
     pub fn set_read_after_close(&mut self, b: bool) -> &mut Self {
         self.read_after_close = b;
+        self.check();
         self
     }
 
     /// Set the max. payload size used when sending data frames. Payloads larger
     /// than the configured max. will be split.
     pub fn set_split_send_size(&mut self, n: usize) -> &mut Self {
         self.split_send_size = n;
+        self.check();
         self
     }
+
+    // TODO: Consider doing the check on creation, not on each builder method call.
+    fn check(&self) {
+        assert!(self.max_stream_receive_window.unwrap_or(u32::MAX) >= DEFAULT_CREDIT);
+        assert!(
+            self.max_connection_receive_window >= self.max_num_streams * DEFAULT_CREDIT as usize
+        );
```

Comment:
Definitely replace these with `Result`s.

Reply:
I moved these into the respective functions, but still panicking. Would you prefer `Result`s over panicking, or were you just opposed to the separate `check` method?